### ISSUE TITLE
Question about finetuning on large datasets (1-50M samples)

### ISSUE THREAD

**Issue by v-dicicco** [meta-llama/llama-recipes#301]
Hello,
I'm using llama-recipes to experiment finetuning of Llama2 on different dataset sizes (spanning 1M to 50M samples and different configuration).
I've noticed a couple of things that make me wonder if the "plain" llama-recipes repo is the right tool to do this kind of experiments, namely:

1. A checkpoint is [evaluated and saved](https://github.com/facebookresearch/llama-recipes/blob/main/src/llama_recipes/utils/train_utils.py#L136) only at the _end of each epoch_
2. The lr_scheduler is [updated](https://github.com/facebookresearch/llama-recipes/blob/3b77a270a87b7a8e4084402bd94017d8b88e5458/src/llama_recipes/utils/train_utils.py#L133) only at the _end of each epoch_

Having large datasets leads to very big epochs, and my fear is that updating too little the lr (for 2.) and evaluating too late (for 1.) the training is far from optimal. Additionally, the datasets listed in the [Dataset page](https://github.com/facebookresearch/llama-recipes/blob/main/docs/Dataset.md) seem to be all in the range of 50-150K samples and this make me wonder:
1. What is the "expected" size the repo is tailored for? 
2. Is it expected to work even with datasets in the range of millions or it is better to customize it (e.g: updating the lr every n steps)? 

Thanks for any hint, hopefully it will help also others! thanks for this great project :)

**Comment by HamidShojanazeri** [comment]
@v-dicicco thanks for opening this discussion, that's correct, we were planning to add checkpointing support on steps rather than epochs, similarly for lr_scheduler. On the dataset size, we haven't tested larger datasets for us but we have customers, reported training with larger dataset using this repo.

I wonder if you are interested to have a PR add the step support for checkpointing and lr_scheduler, would love to review and merge it.

**Comment by wjn1996** [comment]
I have tested this framework, I found that when training for more than 1.5M examples with 'padding' strategies, it should cost more than 6 hours to load data. It's cost too many times before the GPU runs.
